### ISSUE TITLE
Sidekiq 7 incompatibility errors 

### ISSUE THREAD

**Issue by benkruger** [sidekiq-cron/sidekiq-cron#378]
**Gemfile**
```
source 'https://rubygems.org'
ruby '3.1.2'

# Redis
# https://github.com/redis/redis-rb
gem 'redis'

# Async processing
# https://github.com/mperham/sidekiq
gem 'sidekiq'

# https://github.com/ondrejbartas/sidekiq-cron
# Free scheuduled jobs
gem 'sidekiq-cron'
```

**Gemfile.lock**
```
    sidekiq (7.0.1)
      concurrent-ruby (< 2)
      connection_pool (>= 2.3.0)
      rack (>= 2.2.4)
      redis-client (>= 0.9.0)
    sidekiq-cron (1.8.0)
      fugit (~> 1)
      sidekiq (>= 4.2.1)
```

**Error**
```sh
05:20 $  bundle exec sidekiq start
undefined method `options' for Sidekiq:Module

      new_version? ? Sidekiq[key] : Sidekiq.options[key]
                                           ^^^^^^^^
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/options.rb:6:in `[]'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron/schedule_loader.rb:7:in `block in <top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-7.0.1/lib/sidekiq.rb:98:in `configure_server'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron/schedule_loader.rb:6:in `<top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron.rb:4:in `require'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq/cron.rb:4:in `<top (required)>'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq-cron.rb:2:in `require'
/Users/ben/.rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/sidekiq-cron-1.8.0/lib/sidekiq-cron.rb:2:in `<top (required)>'
```

**Comment by markets** [comment]
hey @benkruger 👋🏼 

Sidekiq v7 support is still a wip, if you want to help us, please try it out pointing to this branch: 

```ruby
gem 'sidekiq-cron', github: 'serprex/sidekiq-cron', branch: 'master'
```

More info: https://github.com/sidekiq-cron/sidekiq-cron/pull/369#issuecomment-1301211890

**Comment by benkruger** [comment]
@markets tyvm! This unblocks me and things are working for me again.

**Comment by lisbethw1130** [comment]
We are using serpex/sidekiq-cron with sidekiq 7.0.1, everything works fine, too

**Comment by lisbethw1130** [comment]
@markets hi, we found some issue when task has `date_as_argument: true` or `active_job: true`

```
Sidekiq::Cron::Job.load_from_array([{name: 'test 2', cron: 'every 1 minute', class: 'SomeJob', date_as_argument: true}])
/bundle/ruby/3.1.0/gems/redis-client-0.10.0/lib/redis_client/command_builder.rb:37:in `block in generate': Unsupported command argument type: TrueClass (TypeError)
```

seems it's caused by new redis-client gem used by sidekiq 7

**Comment by markets** [comment]
Hi @lisbethw1130 thanks for your feedback! 

So that error is not related to the Sidekiq v7 update right 🤔? If that's the case, I'd like to move that discussion to another issue, so we can focus on Sidekiq v7 in this one.

**Comment by serprex** [comment]
Sidekiq 6.5 will use redis-client if that gem is available, which the latter error message indicates is the case. Likely there's some code path missing to_s on date_as_argument bool parameter to redis in v7 branch

**Comment by lisbethw1130** [comment]
thanks @serprex , I think I can try to fix it, maybe PR next week

**Comment by lukebooth** [comment]
I'm experiencing an issue with Sidekiq 7. Though this issue was a blocker for me (starting up sidekiq) and it seems like there have been reports of the [WIP PR](https://github.com/sidekiq-cron/sidekiq-cron/pull/369) working fine for some users, so I'm not sure what to think of that.

When [enqueuing jobs](https://github.com/sidekiq-cron/sidekiq-cron/blob/master/lib/sidekiq/cron/poller.rb#L22-L24), with the call to `all`, I think the new adapter is returning an array instead of a hash for the list of jobs.

I needed to convert those arrays into hashes:
```diff
def self.all
  job_hashes = nil
  Sidekiq.redis do |conn|
    set_members = conn.smembers(jobs_key)
    job_hashes = conn.pipelined do |pipeline|
      set_members.each do |key|
        pipeline.hgetall(key)
      end
    end
  end
+ job_hashes = job_hashes.map { |job| job.each_slice(2).to_h } if job_hashes && job_hashes[0].is_a?(Array)
  job_hashes.compact.reject(&:empty?).collect do |h|
    # No need to fetch missing args from Redis since we just got this hash from there
    Sidekiq::Cron::Job.new(h.merge(fetch_missing_args: false))
  end
end
```

Example of what `job_hashes` looked like for me:
```ruby
[["description",
  "",
  "klass",
  "WorkerKlass",
  "status",
  "enabled",
  "symbolize_args",
  "0",
  "name",
  "renew_ssl_certificates",
  "message",
  "{\"retry\":true,\"queue\":\"\ssl\",\"class\":\"WorkerKlass\",\"args\":[]}",
  "active_job",
  "0",
  "queue_name_prefix",
  "",
  "date_as_argument",
  "0",
  "last_enqueue_time",
  "",
  "args",
  "[]",
  "cron",
  "0 1 * * 3",
  "queue_name_delimiter",
  ""],
 ["description",
  "",
  "klass",
  ...
```

To give more context on my situation:
- Using RedisLabs with 6.2.3 database version.
- Because RedisLabs databases currently don't support RESP3, my Sidekiq redis configuration uses `protocol: 2`
- The WIP PR by @serprex (thanks!) I have checked out is https://github.com/sidekiq-cron/sidekiq-cron/commit/8a534b9cd1996eda6d832d9d1cead1ea85a961ed since it's a fluid PR with work being done, might be worth mentioning that.

**Comment by serprex** [comment]
@lukebooth unfortunately redis cloud's redis 6.2 isn't compatible with Sidekiq 7, see https://github.com/mperham/sidekiq/issues/5594

**Comment by lukebooth** [comment]
Ah, gotcha. I've been all over the place seeing if I could get this to work, but never saw that closed issue. Would have saved myself some time 😅. And that comment about getting arrays when Sidekiq expects hashes is exactly what I'm seeing 👍 

> That isn't a proper workaround, many of the response on V2 won't match what Sidekiq expects. e.g. when a hash is expected instead you'll get an Array etc.

Thanks!

**Comment by lisbethw1130** [comment]
seems @serprex fix the argument issues in https://github.com/sidekiq-cron/sidekiq-cron/commit/8a534b9cd1996eda6d832d9d1cead1ea85a961ed
Thanks!
